src/devices/Nrf24l01/Nrf24l01.cs

```diff
@@ -0,0 +1,773 @@
+
+            Write(Command.NRF_W_REGISTER, Register.NRF_TX_ADDR, address);
+
+            _gpio.Write(_ce, PinValue.High);
+        }
+
+        /// <summary>
+        /// Read nRF24L01 Send Address
+        /// </summary>
+        /// <returns>Address</returns>
+        internal Span<byte> ReadTxAddress()
+        {
+            return WriteRead(Command.NRF_R_REGISTER, Register.NRF_TX_ADDR, 5);
+        }
+
+        /// <summary>
+        /// Set Working Channel
+        /// </summary>
+        /// <param name="channel">From 0 to 127</param>
+        internal void SetChannel(byte channel)
+        {
+            if (channel < 0 || channel > 127)
+            {
+                throw new ArgumentOutOfRangeException("Channel needs to be in the range of 0 to 127!");
+            }
+
+            _gpio.Write(_ce, PinValue.Low);
+
+            Write(Command.NRF_W_REGISTER, Register.NRF_RF_CH, channel);
+
+            _gpio.Write(_ce, PinValue.High);
+        }
+
+        /// <summary>
+        /// Read Working Channel
+        /// </summary>
+        /// <returns>Channel</returns>
+        internal byte ReadChannel()
+        {
+            Span<byte> readData = WriteRead(Command.NRF_R_REGISTER, Register.NRF_RF_CH, 1);
+
+            return readData[0];
+        }
+        #endregion
+
+        #region sensor operation
+        internal void Write(Span<byte> writeData)
+        {
+            Span<byte> readBuf = stackalloc byte[writeData.Length];
+
+            _sensor.TransferFullDuplex(writeData, readBuf);
+        }
+
+        internal void Write(Command command, Register register, byte writeByte)
+        {
+            Span<byte> writeBuf = stackalloc byte[2] { (byte)((byte)command + (byte)register), writeByte };
+            Span<byte> readBuf = stackalloc byte[2];
+
+            _sensor.TransferFullDuplex(writeBuf, readBuf);
+        }
+
+        internal void Write(Command command, Register register, Span<byte> writeData)
+        {
+            Span<byte> writeBuf = stackalloc byte[1 + writeData.Length];
+            Span<byte> readBuf = stackalloc byte[1 + writeData.Length];
+
+            writeBuf[0] = (byte)((byte)command + (byte)register);
+            for (int i = 0; i < writeData.Length; i++)
+            {
+                writeBuf[1 + i] = writeData[i];
+            }
+
+            _sensor.TransferFullDuplex(writeBuf, readBuf);
+        }
+
+        internal Span<byte> WriteRead(Command command, Register register, int dataLength)
+        {
+            Span<byte> writeBuf = stackalloc byte[1 + dataLength];
+            Span<byte> readBuf = new byte[1 + dataLength];
+
+            writeBuf[0] = (byte)((byte)command + (byte)register);
+
+            _sensor.TransferFullDuplex(writeBuf, readBuf);
+
+            return readBuf.Slice(1);
+        }
+
+        internal Span<byte> WriteRead(Command command, Register register, Span<byte> writeData)
+        {
+            Span<byte> writeBuf = stackalloc byte[1 + writeData.Length];
```

Comment:
What sort of buffer sizes can we expect here? If they're large they can overflow the stack. If we get large ones in we can rent from `ArrayPool` to reduce allocations and return to the `ArrayPool` at the end of the method.

Solutions for when the inputs are unbounded:

1. Just create a new array (wasteful, but simplest)
2. Always rent the array from the `ArrayPool` (`ArrayPool<byte>.Shared.Rent()`) and return them at the end of the method.
3. Stackalloc if the buffer is small enough (say, under 32 bytes) and otherwise do (1) or (2).

Reply:
Maximum length is 5 bytes😁

Reply:
@ZhangGaoxing in that case please add Debug.Assert directly above :smile: